text/3838-repr-scalable.md

```diff
@@ -0,0 +1,386 @@
+use different registers of sizes to the vectors containing data.
+`repr(scalable)` is used to define vectors containing both data and predicates.
+
+As `repr(scalable(N))` is intended to be a permanently unstable attribute, any
+value of `N` is accepted by the attribute and it is the responsibility of
+whomever is defining the type to provide a valid value. A correct value for `N`
+depends on the purpose of the specific scalable vector type and the
+architecture.
+
+For example, with SVE, the scalable vector register length is a minimum of 128
+bits, must be a multiple of 128 bits and a power of 2; and predicate registers
+have one bit for each byte in the vector registers. So, for `svfloat32_t`
+defined shown above, an `f32` is 32-bits and with `N=4`, the entire minimum
+register length of 128 bits is used (4 x 32 = 128). An intrinsic that takes a
+`svfloat32_t` may also want to accept as an argument a predicate vector with a
+matching four elements (`N=4`), which would only use 4 bits of the predicate
+register rather than the full 16 bits.
+
+See
+[*Manually-chosen or compiler-calculated element count*][manual-or-calculated-element-count]
+for a discussion on why `N` is not calculated by the compiler.
+
+## Properties of scalable vectors
+[properties-of-scalable-vector-types]: #properties-of-scalable-vectors
+
+Scalable vectors are necessarily non-`const Sized` (from [rfcs#3729]) as they
+behave like value types but the exact size cannot be known at compilation time.
+
+[rfcs#3729] allows these types to implement `Clone` (and consequently `Copy`) as
+`Clone` only requires an implementation of `Sized`, irrespective of constness.
+
+Scalable vector types have some further restrictions due to limitations of the
+codegen backend:
+
+- Cannot be stored in compound types (structs, enums, etc) 
+
+    - Including coroutines, so these types cannot be held across an await
+      boundary in async functions
+
+    - `repr(transparent)` newtypes could be permitted with scalable vectors
+
+- Cannot be used in arrays
+
+- Cannot be the type of a static variable.
+
+Some of these limitations may be able to be lifted in future depending on what
+is supported by rustc's codegen backends.
+
+## ABI
+[abi]: #abi
+
+Rust currently always passes SIMD vectors on the stack to avoid ABI mismatches
+between functions annotated with `target_feature` - where the relevant vector
+register is guaranteed to be present - and those without - where the relevant
+vector register might not be present.
+
+However, this approach will not work for scalable vector types as the relevant
+target feature must to be present to use the instruction that can allocate the
+correct size on the stack for the scalable vector.
+
+Therefore, there is an additional restriction that these types cannot be used in
+the argument or return types of functions unless those functions are annotated
+with the relevant target feature.
```

Comment:
Would this not mean that these types cannot be used in function pointers, as we have no knowledge of the target features of a function pointer?

Reply:
For the same reason it would also clash with dyn-compatible traits.

Reply:
I think this is similar to the existing checks we have for `__m256` only being allowed in `extern "C"` functions when AVX is available. They can be used in function ptrs just fine, because someone somewhere must have made the unsafe promise that it is okay to call a `#[target_feature]` function since we actually do have the target feature.

Reply:
Clarified the function pointer point in c6d836de9fca7b4e739478141ffd3a1643fd1ec9 and dyn-compatibility in 530378cf66ce504c386f61001c682dbf199bd4e6.